File: src/Candidato.java
import java.text.NumberFormat;
import java.time.LocalDate;
import java.util.Comparator;
import java.util.Locale;

public class Candidato {

  private String nomeUrna;
  private String numeroCandidato;
  private Boolean participaFederacao;
  private LocalDate dataNascimento;
  private Genero genero;
  private Boolean eleito;
  private int quantidadeVotos;

  private Partido partido;

  public Genero getGenero() {
    return genero;
  }

  public void setGenero(Genero genero) {
    this.genero = genero;
  }

  public String getNomeUrna() {
    return nomeUrna;
  }

  public void setNomeUrna(String nome) {
    this.nomeUrna = nome;
  }

  public String getNumeroCandidato() {
    return numeroCandidato;
  }

  public void setNumeroCandidato(String numeroCandidatoUrna) {
    this.numeroCandidato = numeroCandidatoUrna;
  }

  public Boolean getParticipaFederacao() {
    return this.participaFederacao;
  }

  public void setParticipaFederacao(Boolean participaFederacao) {
    this.participaFederacao = participaFederacao;
  }

  public LocalDate getDataNascimento() {
    return dataNascimento;
  }

  public void setDataNascimento(LocalDate dataNascimento) {
    this.dataNascimento = dataNascimento;
  }

  public Partido getPartido() {
    return partido;
  }

  public void setPartido(Partido partido) {
    this.partido = partido;
  }

  public int getQuantidadeVotos() {
    return quantidadeVotos;
  }

  public void incrementaQuantidadeVotos(int quantidadeVotos) {
    this.quantidadeVotos += quantidadeVotos;
  }

  public Boolean getEleito() {
    return eleito;
  }

  public int getIdade(LocalDate referencia) {
    int idade = referencia.getYear() - dataNascimento.getYear();

    // Caso o aniversario ainda nao ocorreu
    if (referencia.getMonthValue() < dataNascimento.getMonthValue() ||
        (referencia.getMonthValue() == dataNascimento.getMonthValue()
            && referencia.getDayOfMonth() < dataNascimento.getDayOfMonth())) {
      idade--;
    }
    return idade;
  }

  public void setEleito(Boolean eleito) {
    this.eleito = eleito;
  }

  /**
   * @Override
   *           public int compareTo(Candidato outroCandidato) {
   *           int comparaVotos = Integer.compare(outroCandidato.quantidadeVotos,
   *           this.quantidadeVotos);
   *           if (comparaVotos != 0)
   *           return comparaVotos;
   *           return
   *           this.dataNascimento.compareTo(outroCandidato.dataNascimento);
   *           }
   */

  public static class CandidatoComparator implements Comparator<Candidato> {

    private final boolean compararNumeroPartido;

    public CandidatoComparator(boolean compararNumeroPartido) {
      this.compararNumeroPartido = compararNumeroPartido;
    }

    @Override
    public int compare(Candidato c1, Candidato c2) {
      // Comparar votos de forma decrescente
      int comparaVotos = Integer.compare(c2.getQuantidadeVotos(), c1.getQuantidadeVotos());
      if (comparaVotos != 0) {
        return comparaVotos;
      }

      // Caso de empate no número de votos, comparar pelo número partidário (se
      // habilitado)
      if (compararNumeroPartido) {
        int comparaNumeroPartidario = c1.getPartido().getNumero().compareTo(c2.getPartido().getNumero());
        if (comparaNumeroPartidario != 0) {
          return comparaNumeroPartidario;
        }
      }

      // Caso de empate em votos (e número partidário se considerado), comparar pela
      // data de nascimento
      return c1.getDataNascimento().compareTo(c2.getDataNascimento());
    }
  }

  @Override
  public String toString() {
    NumberFormat brFormat = NumberFormat.getInstance(Locale.forLanguageTag("pt-BR"));
    String output = "";
    if (this.partido != null) {
      if (this.participaFederacao)
        output += "*";
      output += this.nomeUrna;
      output += " (" + this.partido.getSigla() + ", " + brFormat.format(this.quantidadeVotos) + " votos)";
    }

    return output;
  }

}

File: implementacao-c++/candidato.cpp
#include "candidato.hpp"

void Candidato::setNomeUrna(const string& nome) {
  this->nomeUrna = nome;
}

string Candidato::getNomeUrna() const {
  return this->nomeUrna;
}

void Candidato::setNumeroCandidato(const string& numero) {
  this->numeroCandidato = numero;
}

string Candidato::getNumeroCandidato() const {
  return this->numeroCandidato;
}

void Candidato::setParticipaFederacao(const bool& participa) {
  this->participaFederacao = participa;
}

bool Candidato::getParticipaFederacao() const {
  return this->participaFederacao;
}

void Candidato::setDataNascimento(const string& dataNascimento) {
  tm data;
  strptime(dataNascimento.c_str(), "%d/%m/%Y", &data);
  this->dataNascimento = data;
}

tm Candidato::getDataNascimento() const {
  return this->dataNascimento;
}

void Candidato::setGenero(const Genero& genero) {
  this->genero = genero;
}

Genero Candidato::getGenero() const {
  return this->genero;
}

void Candidato::setEleito(const bool& eleito) {
  this->eleito = eleito;
}

bool Candidato::getEleito() const {
  return this->eleito;
}

void Candidato::setQuantidadeVotos(const int& votos) {
  this->quantidadeVotos = votos;
}

int Candidato::getQuantidadeVotos() const {
  return this->quantidadeVotos;
}

void Candidato::incrementaVotos(const int& votos) {
  this->quantidadeVotos += votos;
}

int Candidato::getIdade(const tm& dataReferencia) const {
  int idade = dataReferencia.tm_year - this->dataNascimento.tm_year;
  if (dataReferencia.tm_mon < this->dataNascimento.tm_mon || dataReferencia.tm_mday < this->dataNascimento.tm_mday)
    idade--;
  return idade;
}

void Candidato::setPartido(Partido& p) {
  this->partido = &p;
}

Partido* Candidato::getPartido() const {
  return this->partido;
}

bool Candidato::compare(const Candidato& c2, const bool& compararPorNumeroPartido) const {
  // Comparar votos de forma descrescente
  int comparaVotos = this->quantidadeVotos - c2.quantidadeVotos;
  if (comparaVotos != 0)
    return comparaVotos > 0;  // Retorna verdadeiro se c1 tiver menos votos que c2

  // Caso de empate no número de votos, comparar pelo número do partido
  if (compararPorNumeroPartido) {
    int comparaNumeroPartidario = this->getPartido()->getNumero().compare(c2.getPartido()->getNumero());
    if (comparaNumeroPartidario != 0)
      return comparaNumeroPartidario > 0;  // Retorna verdadeiro se c1 tiver número de partido lexicograficamente menor
  }

  // Caso empate, comparar pela data de nascimento
  if (this->getDataNascimento().tm_year != c2.getDataNascimento().tm_year)
    return this->getDataNascimento().tm_year > c2.getDataNascimento().tm_year;  // Comparar anos de nascimento
  if (this->getDataNascimento().tm_mon != c2.getDataNascimento().tm_mon)
    return this->getDataNascimento().tm_mon > c2.getDataNascimento().tm_mon;  // Comparar meses de nascimento
  return this->getDataNascimento().tm_mday > c2.getDataNascimento().tm_mday;  // Comparar dias de nascimento
}
